api/server/httputils/errors.go

```diff
@@ -72,6 +75,104 @@ func GetHTTPErrorStatusCode(err error) int {
 	return statusCode
 }
 
+// FromStatusCode creates an errdef error, based on the provided status-code
+func FromStatusCode(err error, statusCode int) error {
+	if err == nil {
+		return err
+	}
+	switch statusCode {
+	case http.StatusNotFound:
+		err = errdefs.NotFound(err)
+	case http.StatusBadRequest:
+		err = errdefs.InvalidParameter(err)
+	case http.StatusConflict:
+		err = errdefs.Conflict(err)
+	case http.StatusUnauthorized:
+		err = errdefs.Unauthorized(err)
+	case http.StatusServiceUnavailable:
+		err = errdefs.Unavailable(err)
+	case http.StatusForbidden:
+		err = errdefs.Forbidden(err)
+	case http.StatusNotModified:
+		err = errdefs.NotModified(err)
+	case http.StatusNotImplemented:
+		err = errdefs.NotImplemented(err)
+	case http.StatusInternalServerError:
+		if !errdefs.IsSystem(err) && !errdefs.IsUnknown(err) && !errdefs.IsDataLoss(err) && !errdefs.IsDeadline(err) && !errdefs.IsCancelled(err) {
+			err = errdefs.System(err)
+		}
+	default:
+		logrus.WithFields(logrus.Fields{
+			"module":      "api",
+			"status_code": fmt.Sprintf("%d", statusCode),
+		}).Debugf("FIXME: Got an status-code for which error does not match any expected type!!!: %d", statusCode)
+
+		switch {
+		case statusCode >= 200 && statusCode < 400:
+			// it's a client error
```

Comment:
I originally had this at the top to return `nil`, but not sure what's best. Note that the client actually discards these errors; https://github.com/moby/moby/blob/3e5b9cb46688ee5d6886ad3be4cb591f1ba49ed1/client/request.go#L194-L196

Reply:
Related to the above: https://github.com/moby/moby/pull/38462

Reply:
Similar implementation 😄 https://github.com/cpuguy83/strongerrors/blob/67f9d3e5dd563226c4b7956e0c1d8f20ffb026d7/status/http.go#L39-L71


Biggest difference (at a glance);

My implementation does not return `Unknown` for errors if the the status-code is in a known range (i.e.  any unknown 4xx error returns an `InvalidParameter` and any 5xx status return a `System` error.

Downside is that this may make errors Ambiguous; i.e. a `504 Gateway Timeout` (which currently doesn't have a definition / mapping) is transformed into a generic `System` error. _But_ the original status code is preserved, and can be obtained with `err.StatusCode()`